### DISCUSSION TITLE
Magic Link: How can we set an auth cookie on the server before we get to the client?

### DISCUSSION THREAD

**Discussion by fergusmeiklejohn** [supabase/supabase#3762] [Questions]
I'm not using NextJs, we're using Remix https://remix.run as our react framework. The way Remix works, every route has a loader which is called before the route gets to the client.
With Remix it's better practice to avoid useEffect(), instead dealing with data in the server-side Loader functions.
I think with Magic link auth, the DB returns a url with this hash which is then parsed to set the local storage object :
`#access_token=[access_token]&expires_in=3600&refresh_token=[refresh_token]&token_type=bearer&type=[type]"`
I could in theory parse this hash on the server in the Remix Loader function and set the cookie then?
Can we rely on the structure of this hash?
What structure should the cookie have please?
And how might we need to change this to work with OAuth provider authentication?

**Comment by dirathea** [comment]
Just playing around with this, I think we can't parse the hash because it is not sent to the server. I'm not really sure how to, but I tried something like this on the server when the user click the magic link.

```javascript
export let loader: LoaderFunction = async ({request}) => {
  const url = new URL(request.url);
  console.log('hash', url);
};
```

it returns as follows

```
hash URL {
  href: 'http://localhost:3000/',
  origin: 'http://localhost:3000',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost:3000',
  hostname: 'localhost',
  port: '3000',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
```

**Comment by VictorPeralta** [comment]
I've created a starter for Remix with supabase auth both client and serverside, feel free to take a look. 

https://github.com/VictorPeralta/supabase-remix-auth

The basics are, you go through the regular clientside authentication flow, and whenever auth state changes, the server is notified to set or destroy a session cookie which can then be used in loaders and actions.

**Reply by dirathea** [reply]
Thanks for the example @VictorPeralta. I saw that the main diff between your implementation and mine is on the `auth.tsx`, where you handle the form submission, then configure the cookies.

```js
export const action: ActionFunction = async ({ request }) => {
  const formData = await request.formData();
  const authEvent = formData.get("event") as AuthChangeEvent;
  const formSession = formData.get("session");

  if (typeof formSession === "string") {
    const session = JSON.parse(formSession) as Session;
    if (authEvent === "SIGNED_IN") {
      return createUserSession(session.access_token);
    }
    if (authEvent === "SIGNED_OUT") {
      return clearCookie(request);
    }
  }
};
```

some clarification: is the callback from supabase should be something like `localhost:3000/auth`? I just tried it using magic link, and seems like the action function doesn't hit by supabase at all. Is there anything that missing from my implementation?

**Reply by VictorPeralta** [reply]
The callback url shouldn't really matter, the supabase client does its thing, with any authentication method, and uses an auth state change listener that calls the `/auth` endpoint with the created session. If you can create a session in the client and pass it to the server, the cookie should be set.

**Reply by dirathea** [reply]
In short the flow should be something like this :

magic link -> redirect to our page -> onAuthStateChange -> POST to `/auth` to create cookie?

**Reply by parsasi** [reply]
Shouldn't all of this be done on the server (other than initiating OAuth)? It seems counterproductive to use Remix, do all of that work on the client, and then do even more work to sync the server.

**Comment by VictorPeralta** [comment]
That's correct
